src/linalg/linalg_internal_gpu/cuSvd_internal.cu

```diff
@@ -13,216 +10,433 @@ namespace cytnx {
                            boost::intrusive_ptr<Storage_base> &vT,
                            boost::intrusive_ptr<Storage_base> &S, const cytnx_int64 &M,
                            const cytnx_int64 &N) {
-      signed char jobu, jobv;
+      using data_type = cytnx_complex128;
+      cudaDataType cuda_data_type = CUDA_C_64F;
+      cudaDataType cuda_data_typeR = CUDA_R_64F;
+      assert(sizeof(cuDoubleComplex) == sizeof(cytnx_complex128));
 
+      cusolverEigMode_t jobz;
       // if U and vT are NULL ptr, then it will not be computed.
-      jobu = (U->dtype == Type.Void) ? 'N' : 'S';
-      jobv = (vT->dtype == Type.Void) ? 'N' : 'S';
+      jobz = (U->dtype == Type.Void and vT->dtype == Type.Void) ? CUSOLVER_EIG_MODE_NOVECTOR
+                                                                : CUSOLVER_EIG_MODE_VECTOR;
+
+      // const int econ = 0; /* i.e. 'A' in gesvd  */
+      cytnx_int32 econ = 1; /* i.e. 'S' in gesvd  */
 
       // create handles:
       cusolverDnHandle_t cusolverH = NULL;
       checkCudaErrors(cusolverDnCreate(&cusolverH));
 
       cuDoubleComplex *Mij;
-      checkCudaErrors(cudaMalloc((void **)&Mij, M * N * sizeof(cuDoubleComplex)));
+      checkCudaErrors(cudaMalloc((void **)&Mij, M * N * sizeof(data_type)));
       checkCudaErrors(
-        cudaMemcpy(Mij, in->Mem, sizeof(cytnx_complex128) * M * N, cudaMemcpyDeviceToDevice));
-
-      cytnx_int32 min = std::min(M, N);
-      cytnx_int32 ldA = N, ldu = N, ldvT = min;
-      cytnx_int32 lwork = 0;
-
+        cudaMemcpy(Mij, in->Mem, sizeof(data_type) * M * N, cudaMemcpyDeviceToDevice));
+      cytnx_int64 min = std::min(M, N);
+      cytnx_int64 max = std::max(M, N);
+      cytnx_int64 ldA = N, ldu = N, ldvT = M;
+
+      void *UMem, *vTMem;
+      if (U->Mem) {
+        UMem = U->Mem;
+      } else {
+        if (jobz == CUSOLVER_EIG_MODE_VECTOR)
+          checkCudaErrors(cudaMalloc(&UMem, max * max * sizeof(data_type)));
+      }
+      if (vT->Mem) {
+        vTMem = vT->Mem;
+      } else {
+        if (jobz == CUSOLVER_EIG_MODE_VECTOR)
+          checkCudaErrors(cudaMalloc(&vTMem, max * max * sizeof(data_type)));
+      }
+
+      size_t d_lwork = 0; /* size of workspace */
+      void *d_work = nullptr; /* device workspace for getrf */
+      size_t h_lwork = 0; /* size of workspace */
+      void *h_work = nullptr; /* host workspace for getrf */
+      cytnx_double h_err_sigma;
       // query working space :
-      checkCudaErrors(cusolverDnZgesvd_bufferSize(cusolverH, N, M, &lwork));
+      checkCudaErrors(cusolverDnXgesvdp_bufferSize(cusolverH, NULL, /* params */
```

Comment:
I am very skeptical on gesvdp has more accurate result than gesvd. They add perturbation to overcome the non-unitary when corresponding singular value is very small 
[Gesvdp](https://docs.nvidia.com/cuda/cusolver/index.html?highlight=gesvdp#cusolverdnxgesvdp)

Reply:
can you create an issue with this, we need to discuss more on which one to call

Reply:
I thought gesvdp is faster than gesvd, not accurate
However, the gesvd is unable to pass the test, I'll create issue